src/random_bq_to_anndata.py

```diff
@@ -1,46 +1,144 @@
     query = client.query(sql)
     return query.result()
 
 
-def random_bq_to_anndata(project, dataset, num_cells):
+def random_bq_to_anndata(project, dataset, num_cells, output_file_prefix):
     client = bigquery.Client(project=project)
-    cell_data = get_cell_data(project, dataset, client, num_cells)
-    adata = ad.AnnData()
-    for row in list(cell_data):
-        # insert VS-507 here
-        pass
 
+    print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
+    random_cell_ids = get_random_cell_ids(project, dataset, client, num_cells)
+
+    # Read the cell information and store for later
+    cells = get_cells(project, dataset, client, random_cell_ids)
+    original_cell_ids = []
+    cell_types = []
+    cas_cell_index_to_row_num = {}
+    row_num = 0
+    for cell in cells:
+        original_cell_ids.append(cell.original_cell_id)
+        cell_types.append(cell.cell_type)
+        cas_cell_index_to_row_num[cell.cas_cell_index] = row_num
+        row_num += 1
+
+    # Read the feature information and store for later.
+    features = get_features(project, dataset, client)
+
+    feature_ids = []
+    feature_names = []
+    cas_feature_index_to_col_num = {}
+    col_num = 0
+    for feature in features:
+        feature_ids.append(feature.original_feature_id)
+        feature_names.append(feature.feature_name)
+        cas_feature_index_to_col_num[feature.cas_feature_index] = col_num
+        col_num += 1
+
+    # Note that this method requires that the result set returned by get_cell_data be sorted by cas_cell_index (or, could also be sorted by original_cell_id)
+    cell_data = get_matrix_data(project, dataset, client, random_cell_ids)
+
+    (rows, columns, data) = generate_coo_sparse_matrix(cell_data, cas_cell_index_to_row_num, cas_feature_index_to_col_num)
+    # Create the matrix from the sparse data representation generated above.
+    counts = coo_matrix((data, (rows, columns)), shape=(len(cells), len(features)), dtype=np.float32)
+
+    # Had to convert the COO matrix to CSR for loading into AnnData
+    adata = ad.AnnData(counts.tocsr(copy=False))
+    adata.obs_names = original_cell_ids
+    adata.obs["cell_type"] = cell_types
+    adata.var_names = feature_ids
+    adata.var["feature_name"] = feature_names
+
+    # See https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.raw.html?highlight=raw#anndata.AnnData.raw
+    #  for why we set 'raw' thusly: "The raw attribute is initialized with the current content of an object by setting:"
+    adata.raw = adata
+    adata.write(f'{output_file_prefix}.h5ad', compression="gzip")
+
+def generate_coo_sparse_matrix(cell_data, cas_cell_index_to_row_num, cas_feature_index_to_col_num):
```

Comment:
naming nitpick, but this doesn't actually include generating the coo sparse matrix

Reply:
Yup, I've changed the name.